### PR TITLE
Adding an Item container prop to enclose the items in the list.

### PR DESCRIPTION
**Pull request by arunreddy-flipkart**
The idea is to write a native component implementing an algorithm to remove the jitter by finding the overlaps between the subviews being rendered and shifting them accordingly. The native component is currently ready for android and under development for IOS. The user can pass two props to the RLV namely itemContainer and itemContainerProps. The itemContainer is the component while itemContainerProps is an object consisting of the custom props to be passed to the component. If the itemContainer prop is provided, it will be used to enclose the list and a shifting algorithm will run at the native level to remove the jitter otherwise the flow will be as before.

### CODE REVIEW

File: src/core/RecyclerListView.tsx [hunk 1]
```diff
@@ -151,7 +153,7 @@ export default class RecyclerListView<P extends RecyclerListViewProps, S extends
     private _cachedLayouts?: Layout[];
     private _scrollComponent: BaseScrollComponent | null = null;
```
Comment (naqvitalha) [inline src/core/RecyclerListView.tsx]: mention the reason in comment

File: src/core/ItemContainer.tsx [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import * as React from "react";
+import { ComponentCompat } from "../utils/ComponentCompat";
+
+export interface ItemContainerProps {
+    horizontal: boolean;
+    layoutOffset: number;
+    layoutInterval: number;
```
Comment (naqvitalha) [inline src/core/ItemContainer.tsx]: change the name

File: src/core/ItemContainer.tsx [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import * as React from "react";
+import { ComponentCompat } from "../utils/ComponentCompat";
+
+export interface ItemContainerProps {
```
Comment (naqvitalha) [inline src/core/ItemContainer.tsx]: call it content container

File: src/core/RecyclerListView.tsx [hunk 4]
```diff
@@ -363,7 +369,13 @@ export default class RecyclerListView<P extends RecyclerListViewProps, S extends
                 onScroll={this._onScroll}
                 onSizeChanged={this._onSizeChanged}
                 contentHeight={this._initComplete ? this._virtualRenderer.getLayoutDimension().height : 0}
-                contentWidth={this._initComplete ? this._virtualRenderer.getLayoutDimension().width : 0}>
+                contentWidth={this._initComplete ? this._virtualRenderer.getLayoutDimension().width : 0}
+                itemContainerProps={this.props.itemContainer ? {
+                    horizontal : this.props.isHorizontal,
+                    layoutOffset : this.getCurrentScrollOffset(),
+                    layoutInterval: windowBound + this.getCurrentRenderAheadOffset(),
```
Comment (naqvitalha) [inline src/core/RecyclerListView.tsx]: all of need here is renderAheadOffset

File: src/platform/reactnative/scrollcomponent/ScrollComponent.tsx [hunk 5]
```diff
@@ -64,12 +65,24 @@ export default class ScrollComponent extends BaseScrollComponent {
...
                 <View style={{ flexDirection: this.props.isHorizontal ? "row" : "column" }}>
-                    <View style={{
-                        height: this.props.contentHeight,
-                        width: this.props.contentWidth,
-                    }}>
-                        {this.props.children}
-                    </View>
+                    {
+                        (!this.props.itemAnimator && this.props.itemContainer) ? (
+                            <ItemContainer style={{
```
Comment (naqvitalha) [inline src/platform/reactnative/scrollcomponent/ScrollComponent.tsx]: This big check is not required now

File: src/core/RecyclerListView.tsx (hunk 1 above)
Comment (arunreddy-flipkart) [reply src/core/RecyclerListView.tsx]: Comment has been added explaining the reason why baseItemAnimator is used

File: src/core/ItemContainer.tsx (hunk 2 above)
Comment (arunreddy-flipkart) [reply src/core/ItemContainer.tsx]: The prop names have been changed to scrollOffset and windowSize

File: src/core/ItemContainer.tsx (hunk 3 above)
Comment (arunreddy-flipkart) [reply src/core/ItemContainer.tsx]: This file has been removed as it was not necessary

File: src/core/RecyclerListView.tsx (hunk 4 above)
Comment (arunreddy-flipkart) [reply src/core/RecyclerListView.tsx]: Recyclerlistview has to pass these props to scrollcomponent as only RLV will have the correct values such as that of scrollOffset

File: src/platform/reactnative/scrollcomponent/ScrollComponent.tsx (hunk 5 above)
Comment (arunreddy-flipkart) [reply src/platform/reactnative/scrollcomponent/ScrollComponent.tsx]: Has been refactored to have a contentContainer function

File: src/core/RecyclerListView.tsx [hunk 6]
```diff
@@ -104,7 +104,7 @@ export interface RecyclerListViewProps {
     optimizeForInsertDeleteAnimations?: boolean;
     style?: object | number;
     debugHandlers?: DebugHandlers;
-
+    contentContainer?: (props?: object, children?: React.ReactNode) => JSX.Element | null;
```
Comment (naqvitalha) [inline src/core/RecyclerListView.tsx]: return react.reactnode

File: src/core/RecyclerListView.tsx [hunk 7]
```diff
@@ -151,7 +151,10 @@ export default class RecyclerListView<P extends RecyclerListViewProps, S extends
     private _cachedLayouts?: Layout[];
     private _scrollComponent: BaseScrollComponent | null = null;
 
-    private _defaultItemAnimator: ItemAnimator = new DefaultItemAnimator();
+    //If the native content container is used, then positions of the list items are changed on the native side. The animated library used
+    //by the default item animator also changes the same positions which could lead to inconsistency. Hence, the base item animator which
+    //does not perform any such animations will be used.
+    private _defaultItemAnimator: ItemAnimator = this.props.contentContainer ? new BaseItemAnimator() : new DefaultItemAnimator();
```
Comment (naqvitalha) [inline src/core/RecyclerListView.tsx]: Make base as default

File: src/core/RecyclerListView.tsx (hunk 6 above)
Comment (naqvitalha) [reply src/core/RecyclerListView.tsx]: change the name to renderContentContainer

File: src/core/RecyclerListView.tsx [hunk 8]
```diff
@@ -336,6 +339,12 @@ export default class RecyclerListView<P extends RecyclerListViewProps, S extends
     }
 
     public renderCompat(): JSX.Element {
+        const windowBound = this.props.isHorizontal ? this.getRenderedSize().width : this.getRenderedSize().height;
+        const contentContainerProps = {
```
Comment (naqvitalha) [inline src/core/RecyclerListView.tsx]: move to scroll component

File: src/core/RecyclerListView.tsx (hunk 6 above)
Comment (arunreddy-flipkart) [reply src/core/RecyclerListView.tsx]: Done

File: src/core/RecyclerListView.tsx (hunk 7 above)
Comment (arunreddy-flipkart) [reply src/core/RecyclerListView.tsx]: Done

File: src/core/RecyclerListView.tsx (hunk 8 above)
Comment (arunreddy-flipkart) [reply src/core/RecyclerListView.tsx]: Done

**Pull request merged**